overlord/hookstate/hookmgr.go

```diff
@@ -351,23 +339,39 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
 		close(hookCompleted)
 	}()
 
-	select {
-	// Hook completed; it may or may not have been successful.
-	case <-hookCompleted:
-		return buffer.Bytes(), hookError
-	// max timeout reached
-	case <-killTimerCh:
-		if err := killemAll(command); err != nil {
-			return nil, fmt.Errorf("cannot abort hook %q: %s", hookName, err)
-		}
-		// its ok to use buffer.Bytes() here, killemAll() calls
-		// cmd.Wait() internally and errors if that does not work
-		return buffer.Bytes(), fmt.Errorf("exceeded maximum runtime of %s", maxRuntime)
+	var abortOrTimeoutError error
+out:
+	for {
+		select {
+		// Hook completed; it may or may not have been successful.
+		case <-hookCompleted:
+			break out
+		// max timeout reached
+		case <-killTimerCh:
+			if err := killemAll(command); err != nil {
+				return nil, fmt.Errorf("cannot abort hook %q: %s", hookName, err)
+			}
+			abortOrTimeoutError = fmt.Errorf("exceeded maximum runtime of %s", maxRuntime)
+		// cmdWaitTimeout was reached, i.e. command.Wait() did not
+		// finish in a reasonable amount of time, we can not use
+		// buffer in this case
+		case <-cmdWaitTimerCh:
+			return nil, fmt.Errorf("exceeded maximum runtime of %s and did not stop", maxRuntime)
 		// Hook was aborted.
-	case <-tomb.Dying():
-		if err := killemAll(command); err != nil {
-			return nil, fmt.Errorf("cannot abort hook %q: %s", hookName, err)
+		case <-tomb.Dying():
+			// tomb.Dyning() will keep sending but we just need
+			// to handle it once
+			if abortOrTimeoutError != nil {
+				continue
```

Comment:
this might busy loop,  I think you could:

    dyingCh := tomb.Dying()

before the select, and then

    dyingCh = nil

once you have seen it trigger once